File: 560-subarray-sum-equals-k/subarray-sum-equals-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    const prefixSums = new Map();

    prefixSums.set(0, 1);

    let currSum = 0;
    let totalSubarrays = 0;

    for (let i = 0; i < nums.length; i++) {
        currSum += nums[i];

        const complement = currSum - k;
        if (prefixSums.has(complement)) {
            totalSubarrays += prefixSums.get(complement);
        }

        const count = prefixSums.get(currSum) || 0;
        prefixSums.set(currSum, count + 1);
    }

    return totalSubarrays;
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        # Map to store the frequency of prefix sums
        # Initialize with {0: 1} to handle subarrays starting from index 0
        prefix_sums = {0: 1}

        curr_sum = 0
        total_subarrays = 0

        for num in nums:
            # Update the running cumulative sum
            curr_sum += num

            # Check if the complement (curr_sum - k) exists in our history
            complement = curr_sum - k
            if complement in prefix_sums:
                total_subarrays += prefix_sums[complement]

            # Record the current prefix sum in the hash map
            prefix_sums[curr_sum] = prefix_sums.get(curr_sum, 0) + 1

        return total_subarrays